File: leetcode/solutions/0049.GroupAnagrams/groupAnagrams.go
package solution

func groupAnagrams(strs []string) [][]string {
	const first = 'a'
	tmp := make(map[[26]int][]string)

	for _, str := range strs {
		key := [26]int{}

		for _, r := range str {
			key[r-first]++
		}

		tmp[key] = append(tmp[key], str)
	}

	result := [][]string{}

	for _, strs := range tmp {
		result = append(result, strs)
	}

	return result
}

File: leetcode/solutions/0049.GroupAnagrams/groupAnagrams.ts
export default function groupAnagrams(strs: string[]): string[][] {
  const map: Map<string, string[]> = new Map();

  for (const str of strs) {
    const key = [...str].sort().join("");
    const strs = map.get(key) || [];
    strs.push(str);
    map.set(key, strs);
  }

  return [...map.values()];
}
